README.md
# Logstash Javascript Filter

[![Travis Build Status](https://travis-ci.com/kares/logstash-filter-javascript.svg)](https://travis-ci.com/kares/logstash-filter-javascript)

This is a plugin for [Logstash](https://www.elastic.co/guide/en/logstash/current/introduction.html), compatible with LS 
versions **>= 6.8**, that allows you to interact with pipeline events using scripts written in Javascript. 
It works in a similar way as Logstash's (official) [Ruby filter](https://www.elastic.co/guide/en/logstash/current/plugins-filters-ruby.html). 

**Only Java 8 - 13 is supported**, running LS on Java 15 or later won't work due the removal of the Nashorn Javascript engine.

**DISCLAIMER: Plugin is considered a (working) experiment and is no way as battle tested as (official) Logstash plugins supported 
by [Elastic](https://www.elastic.co/support/matrix#matrix_logstash_plugins).** 

In general, performance wise, you can expect the same throughput as with the Ruby filter.
Also, performance of the Nashorn Javascript engine *might* vary between Java versions.

## Usage

To inline Javascript in your filter, place all code in the `code` option. This code will be executed for every event the 
filter receives. For example, to cancel 90% of events, you can do this:
```javascript
  filter {
    javascript {
      code => "if (java.lang.Math.random() <= 0.9) event.cancel()"
    }
  }
```

You can also place JS code in the `init` option - it will be executed only once during the plugin's initialization phase.
This is a great place to "feature validate" the Javascript engine:

```javascript
  filter {
    javascript {
      init => "if (Number.MIN_VALUE <= 0) throw new Error(0); if (parseInt('f*ck', 16) !== 15) throw 'f*ck'"
      code => 'event.setField("message", "b" + "a" + +"a" + "a")'
    }
  }
```

### Installing

`$LS_HOME/bin/logstash-plugin install logstash-filter-javascript`

### Configuration

TO-BE-CONTINUED...

### Differences from Ruby filter

Unlike the Ruby filter, which allows you to hook into the LS execution runtime, the Javascript filter starts an isolated
JS engine on every filter use.

There's no `new_event_block` callback hook implemented in the Javascript filter, this one (if requested) deserves more 
thought as it just felt a bit "hacky" to copy what the Ruby filter does. 

The Javascript filter does not expose a `register` function (with `script_params`), instead you can use `init_parameters` 
to set variables in the global scope which will than be accessible from within the `filter` function. 

### Tips & Tricks

Nashorn defaults to ECMAScript 5.1 by default which lacks the compact arrow `arg => ...` syntax or the `let` keyword.
There's (incomplete) support for ECMA 6 but requires setting a system properly, navigate to *config/jvm.options* and add :
```
-Dnashorn.args=--language=es6
```

Be aware of scripting Java types with Nashorn as not all native Javascript APIs will handle those seamlessly and 
might lead to surprising results e.g.

```javascript
var json = JSON.stringify(event.toMap()); // undefined
// as JSON does not handle a java.util.Map returned from the LS event

// one can instead convert Java types to native JS objects e.g.
var map = event.toMap()
var obj = {}
for each (var key in map.keySet()) obj[key] = map.get(key) // Nashorn for-each extension for Java arrays/collections
```

You can set plain-old Javascript objects as values on the event, LS will see them as maps and convert them accordingly:

```javascript
var obj = { foo: "bar", truthy: true, aNull: null, number: 11.1 }
event.setField('js.values', obj)
// be aware when JS values contain function types as they might lead to issues
```

```javascript
// using JS types with a Java type system might lead to issues e.g. setting it on an event e.g.
event.setField('unexpected-value', undefined); // LS will complain not being able to handle :
// Missing Converter handling for full class name=jdk.nashorn.internal.runtime.Undefined
```

There's no `console.log` with Nashorn, however you could use Java's system output for debugging purposes :
```javascript
function puts(msg) {
  java.lang.System.out.println(msg)  
}

puts('event: ' + event.toMap());

// or simply the built-in print method :
print('event: ', event)
```
**NOTE**: be aware to remove such debugging statements in production to not fill up LS' standard output!

## Developing

### 1. Plugin Development and Testing

#### Code

- To get started, you'll need JRuby (>= 9.1) with the Bundler gem installed.

- Install dependencies
```sh
jruby -S bundle
```

#### Test

```sh
jruby -rbundler/setup -S rspec
```

### 2. Running your unpublished Plugin in Logstash

- Edit Logstash's `Gemfile` and add the local plugin path e.g.:
```ruby
gem "logstash-filter-javascript", :path => "path/to/local/logstash-filter-javascript"
```
- Install plugin
```sh
bin/logstash-plugin install --no-verify
```
- Run Logstash with your plugin
```sh
bin/logstash -e "filter { javascript { code => \"print('Hello from JS: ' + event.getField('message'))\" } }"
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-filter-awesome.gemspec
```
- Install the plugin from the Logstash home
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Start Logstash and proceed to test the plugin

## Copyright

(c) 2020 [Karol Bucek](https://github.com/kares).
See LICENSE (http://www.apache.org/licenses/LICENSE-2.0) for details.


CHANGELOG.md
## 0.1.0
 - inception! :collision: :star:



spec/filters/javascript_spec.rb
# encoding: utf-8
require_relative '../spec_helper'
require "logstash/filters/javascript"
require "logstash/filters/date"

describe LogStash::Filters::Javascript do

  let(:event) do
    LogStash::Event.new "message" => "hello javascript"
  end

  subject(:plugin) { ::LogStash::Filters::Javascript.new(options) }

  let(:options) { fail 'let(:options) needs to be defined' }

  context "inline (code) script" do

    describe "using Java event API" do
      config <<-CONFIG
        filter {
          date {
            match => [ "date", "ISO8601" ]
            locale => "en"
            timezone => "UTC"
          }
          javascript {
            code => "var map = event.toMap(); var obj = {}; for each (var key in map.keySet()) obj[key] = map.get(key); event.setField('js.obj', obj)"
          }
        }
      CONFIG

      sample("message" => "hello", "date" => "2014-09-23T00:00:00-0800", "truthy" => true, "number" => 1.11) do
        js_obj = subject.get('js.obj') # LS event auto-converts map-like objects
        expect( js_obj['message'] ).to eql 'hello'
        expect( js_obj['number'] ).to eql 1.11
        expect( js_obj['truthy'] ).to eql true
      end
    end

    describe "returning new event" do
      let(:options) { { 'code' => "return event.clone()" } }
      before(:each) { plugin.register }

      it "proceeds with new event and cancels old" do
        events = plugin.multi_filter [ event ]
        expect( events.length ).to eq 2
        expect( event.cancelled? ).to be true
      end
    end

    describe "returning js object" do
      let(:options) { { 'code' => 'return { foo: {}, message: event.getField("message") + "!" }' } }
      before(:each) { plugin.register }

      it "creates new event and cancels old" do
        events = plugin.multi_filter [ event ]
        expect( events.length ).to eq 2
        expect( events[0] ).to be event
        expect( event.cancelled? ).to be true

        expect( events[1].get('message') ).to eql "hello javascript!"
        expect( events[1].get('foo') ).to eql Hash.new
      end
    end

    describe "returning same event" do
      let(:options) { { 'code' => "return event" } }
      before(:each) { plugin.register }

      it "does not cancel event" do
        events = plugin.multi_filter [ event ]
        expect( events.length ).to eq 1
        expect( events[0] ).to equal(event)
        expect( event.cancelled? ).to be false
      end
    end

    describe "returning multiple events" do
      let(:options) { { 'code' => "return [event.clone(), event.clone(), event.clone()]" } }
      before(:each) { plugin.register }

      it "produces more events" do
        expect { |block| plugin.filter(event, &block) }.to yield_control.exactly(3).times
      end
    end

    # describe "catch all JS exceptions" do
    #   config <<-CONFIG
    #     filter {
    #       javascript {
    #         code => "throw 42"
    #       }
    #     }
    #   CONFIG
    #
    #   sample("message" => "hello", "date" => "2014-09-23T00:00:00-0800") do
    #     expect( subject.get("tags") ).to eql ["_javascriptexception"]
    #   end
    # end

    describe "throwing error" do
      let(:options) { { 'code' => "if (true) throw new Error('a message')" } }
      before(:each) { plugin.register }

      it "handles Error" do
        expect( plugin.logger ).to receive(:error).
            with('could not process event due:', hash_including(
                :javascript_error => 'Error: a message',
                :javascript_line => 1 + 1, :javascript_column => 10)
            ) #.and_call_original

        events = plugin.multi_filter [ event ]
        expect( events.length ).to eq 1
        expect( events[0] ).to equal(event)
        expect( event.get('tags') ).to eql ["_javascriptexception"]
      end
    end

    describe "throwing object" do
      let(:options) { { 'code' => "\n throw 42" } }
      before(:each) { plugin.register }

      it "handles (javascript) error" do
        expect( plugin.logger ).to receive(:error).
            with('could not process event due:', hash_including(
                :javascript_error => 42,
                :javascript_line => 2 + 1, :javascript_column => 1)
            ) #.and_call_original

        events = plugin.multi_filter [ event ]
        expect( events.length ).to eq 1
        expect( events[0] ).to equal(event)
        expect( event.get('tags') ).to eql ["_javascriptexception"]
      end
    end

    describe "unexpected return value" do
      let(:options) { { 'code' => "if (true) return false" } }

      it "raises script error" do
        plugin.register
        expect { plugin.multi_filter [ event ] }.to raise_error(LogStash::Filters::Javascript::ScriptError)
      end
    end

    describe "invalid script" do
      let(:options) { { 'code' => "if (true) { 'okay' }\nif (false) invalid syntax {;" } }
      before(:each) do
        expect( plugin.logger ).to receive(:error)
      end

      it "should error out during register" do
        expect { plugin.register }.to raise_error(Java::JavaxScript::ScriptException)
      end

      it "reports correct error line" do
        begin
          plugin.register
          fail('syntax error expected')
        rescue Java::JavaxScript::ScriptException => e
          expect( e.cause.message ).to include "Expected ; but found syntax\nif (false) invalid syntax {;"
        end
      end
    end

  end

  context "path option" do
    let(:options) { { 'path' => "spec/fixtures/throwIfErrorFieldSet.js" } }
    before(:each) { plugin.register }

    it "works normally (wout return)" do
      plugin.multi_filter [ event ]
      expect( event.get('filtered') ).to be true
    end

    it "raises error" do
      event.set('error', 'FROM SPEC')
      plugin.multi_filter [ event ]
      expect( event.get('tags') ).to eql ["_javascriptexception"]
    end
  end

  context "non-readable path option" do
    let(:options) { { 'path' => "__INVALID__.js" } }

    it "raises configuration error" do
      expect { plugin.register }.to raise_error(LogStash::ConfigurationError)
    end
  end

  context "multiple path(s)" do
    let(:options) { {
        'path' => [ 'spec/fixtures/underscore.js', 'spec/fixtures/ramda.js' ],
        'code' => 'event.setField( "last", _.last([0, 1, 2]) ); ' +
                  'event.setField( "add", R.add(10)(1) ); '
    } }
    before(:each) { plugin.register }

    it "executes code using _" do
      plugin.multi_filter [ event ]
      expect( event.get('last') ).to eql 2
      expect( event.get('add') ).to eql 11
    end

  end

  context "init script" do
    let(:options) { { 'init' => 'var foo = "bar"', 'code' => "event.setField('foo', foo)" } }

    it "sets a variable in 'global' scope" do
      plugin.register
      plugin.multi_filter [ event ]
      expect( event.get('foo') ).to eql 'bar'
    end

    context 'wtfjs' do
      let(:options) { {
          'init' => "if (Number.MIN_VALUE <= 0) throw new Error(0); if (parseInt('f*ck', 16) !== 15) throw 'f*ck'",
          'code' => 'event.setField("message", "b" + "a" + +"a" + "a")'
      } }

      it 'is still javascript' do
        plugin.register
        plugin.multi_filter [ event ]
        expect( event.get('message') ).to eql 'baNaNa'
      end
    end
  end

  context "init parameters" do
    let(:options) { {
        'init_parameters' => { 'env' => 'debug', 'fieldName' => 'x', 'multiplier' => 10 },
        'path' => "spec/fixtures/field_multiplier.js"
    } }

    it "calculates x * multiplier" do
      plugin.register
      event.set('x', 4.2)
      plugin.multi_filter [ event ]
      expect( event.get('x') ).to eql 42.0
    end
  end

end


spec/fixtures/field_multiplier.js
// fieldName
// multiplier

function filter(event) {
  var value = Number( event.getField(fieldName) ) * multiplier
  event.setField(fieldName, value)
  if (env === 'debug') print("calculated value: " + value + " for " + event);
  // Filter blocks must return any events that are to be passed on
  // return a nil or [] here if all events are to be cancelled
  // You can even return one or more brand new events here!
  return [ event ]
}

/*
test "standard flow" do
  parameters do
    { "field" => "myfield", "multiplier" => 3 }
  end

  in_event { { "myfield" => 123 } }

  expect("there to be only one result event") do |events|
    events.size == 1
  end

  expect("result to be equal to 123*3(369)") do |events|
    events.first.get("myfield") == 369
  end
end
*/

spec/fixtures/throwIfErrorFieldSet.js

function filter(event) {
  var error = event.getField('error');
  if (error) throw new TypeError(error)
  event.setField('filtered', true)
}


lib/logstash/filters/javascript.rb
# encoding: utf-8
require "logstash/filters/base"
require "logstash/namespace"

# Execute JS code.
#
# For example, to cancel 90% of events, you can do this:
# [source,ruby]
#     filter {
#       javascript {
#         # Cancel 90% of events
#         code => "if (java.lang.Math.random() <= 0.9) event.cancel()"
#       }
#     }
module LogStash module Filters class Javascript < Base

  java_import 'java.util.Map'
  java_import 'jdk.nashorn.api.scripting.NashornException'
  java_import 'jdk.nashorn.api.scripting.ScriptObjectMirror'
  JEvent = org.logstash.Event
  private_constant :JEvent

  class ScriptError < StandardError; end

  config_name "javascript"

  # Any code to execute at plugin startup-time
  config :init, :validate => :string

  # Parameters for this specific script, these are set as globals.
  config :init_parameters, :type => :hash, :default => {}

  # The code to execute for every event.
  # You will have an `event` variable available that is the event itself.
  config :code, :validate => :string

  # Path to the script.js
  config :path, :validate => :path, :list => true

  # Tag to add to events that cause an exception in the script filter
  config :tag_on_exception, :type => :string, :default => "_javascriptexception"

  # Flag for add exception message to tag_on_exception
  #config :tag_with_exception_message, :type => :boolean, :default => false

  def initialize(*params)
    super(*params)
    @script = Script.new(nil, init_parameters, logger)
    @script.js_eval @init if @init
    @js_filter = nil
  end

  def register
    @js_filter = nil # TODO: devutils' sample helper causes plugin.register to happen twice

    if @path
      @path.each { |path| @script.js_eval(::File.read(path), path: path) }
      @js_filter = @script.get('filter') # `expecting a `function filter(event) {}`
      if @js_filter.nil?
        raise ScriptError, "script at '#{@path}' does not define a filter(event) function" if @code.nil?
      else
        if @js_filter.is_a?(ScriptObjectMirror)
          unless @js_filter.isFunction
            raise ScriptError, "script at '#{@path}' defines a 'filter' property that isn't a function (got type: #{@js_filter.getClassName})"
          end
        else
          raise ScriptError, "script at '#{@path}' defines a 'filter' property that isn't a function (got value: #{@js_filter.inspect})"
        end
      end
    end

    if @code
      if @js_filter
        msg = "Script(s) provided by 'path' defined a filter function while 'code' => ... is also set, this is ambiguous!"
        @logger.error(msg); raise LogStash::ConfigurationError.new(msg)
      end

      @js_filter = @script.js_eval(@code) { "(function filter(event) {\n#{@code} } )" } # jdk.nashorn.api.scripting.JSObject
    end

    @script.verify
  end

  def filter(event, &block)
    java_event = event.to_java
    begin
      js_return = @script.js_call(@js_filter, java_event)
      filter_matched(event)
    rescue => e
      @logger.error("could not process event due:", error_details(e))
      tag_exception(event, e)
      return event
    end
    event.cancel unless filter_results(java_event, js_return, &block)
  end

  private

  # @param results (JS array) in a `jdk.nashorn.api.scripting.ScriptObjectMirror`
  def filter_results(event, js_return)
    if js_return.nil? # explicit `return null`
      # drop event (return false)
    elsif ScriptObjectMirror.isUndefined(js_return) # jdk.nashorn.internal.runtime.Undefined
      return true # do not drop (assume it's been dealt with e.g. `event.cancel()`)
    elsif js_return.is_a?(ScriptObjectMirror)
      if js_return.isArray
        i = 0; returned_original = false
        while i < js_return.size
          evt = js_return.getSlot(i)
          if event.equal?(evt)
            returned_original = true
          else
            yield wrap_event(evt)
          end
          i += 1
        end
        return returned_original
      else
        begin
          evt = wrap_event(js_return) # JSObject implement Map interface
        rescue => e
          raise e # TODO we should attempt to provide a better exception here if we can not convert the JS (map) object
          # raise ScriptError, "javascript did not return an event/array (or null) from 'filter', got: #{js_return.getClassName}"
        else
          yield evt
        end
      end
    elsif js_return.is_a?(JEvent) || js_return.is_a?(Map)
      return true if event.equal?(js_return)
      yield wrap_event(js_return)
    else
      raise ScriptError, "javascript did not return an event/array (or null) from 'filter', got: #{js_return.inspect}"
    end
    false # script did not return original event
  end

  # NOTE: JS code is expected to work with Java event API
  # @param js_event a org.logstash.Event or simply a plain-old (map-like) JS object
  def wrap_event(js_event)
    js_event = JEvent.new(js_event) unless js_event.is_a?(JEvent)
    Event.new(js_event)
  end

  def tag_exception(event, e)
    if @tag_with_exception_message && e.message
      event.tag("#{@tag_on_exception}: #{e.message}")
    end
    event.tag(@tag_on_exception) if @tag_on_exception
  end

  def error_details(e)
    details = { :exception => e.class, :message => e.message }
    if e.is_a?(NashornException)
      details[:message] = e.toString
      details[:backtrace] = e.backtrace if logger.debug?
      js_error = e.getEcmaError
      js_error = js_error.toString if js_error.is_a?(ScriptObjectMirror)
      details[:javascript_error] = js_error
      details[:javascript_file] = e.getFileName # '<eval>' for incline code
      details[:javascript_line] = e.getLineNumber # inline code lines are +1
      details[:javascript_column] = e.getColumnNumber
      details[:javascript_trace] = NashornException.getScriptFrames(e)
    else
      details[:backtrace] = e.backtrace
    end
    details
  end

  class Script

    FILENAME = javax.script.ScriptEngine::FILENAME

    attr_reader :context

    # @param context the JS this context for the filter function
    # @param params additional JS (key-value) parameters to set 'globally'
    def initialize(context, params, logger)
      @logger = logger
      @engine = javax.script.ScriptEngineManager.new.getEngineByName("nashorn")
      @context = context

      factory = @engine.getFactory
      logger.debug "initialized javascript (#{factory.getLanguageVersion}) engine:", name: factory.getEngineName, version: factory.getEngineVersion

      params.each { |name, value| @engine.put(name, value) }
    end

    def js_eval(code, path: nil)
      filename = @engine.get(FILENAME)
      @engine.put(FILENAME, path)
      @engine.eval block_given? ? yield : code
    rescue => e # (non-checked) Java::JavaxScript::ScriptException, e.g.
      # Java::JavaxScript::ScriptException (TypeError: Cannot read property "far" from undefined in <eval> at line number 2)
      @logger.error "failed to evaluate javascript code:", code_hint(code, path).merge(message: e.message)
      raise e
    ensure
      @engine.put(FILENAME, filename)
    end

    def js_call(js_fn, arg)
      js_fn.call(@context, arg)
    end

    # @return nil if no such property
    def get(name)
      @engine.get(name)
    end

    def verify
      true # NOTE can we do more JS code checks?
    end

    private

    MAX_LINE_LENGTH = 50

    def code_hint(code, path)
      return { path: path } if path
      lines = code.split("\n")
      code = lines.find { |line| ! line.strip.empty? }
      if code.length <= MAX_LINE_LENGTH
        code = "#{code}..." if lines.size > 1
      else
        code = "#{code[0, MAX_LINE_LENGTH - 3]}..."
      end
      { code: code }
    end

  end

end end end
